### PR TITLE
Rename RunConfig to ModelRunConfig and create new RunConfig

### PR DESCRIPTION
**Pull request by tgerdesnv**
(And do the same for the corresponding generators)

### CODE REVIEW

File: model_analyzer/config/run/model_run_config.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        Returns true if the run_config is valid and should be run. Else false
+        """
+        model_config = self._model_config.get_config()
+
+        max_batch_size = model_config[
+            'max_batch_size'] if 'max_batch_size' in model_config else 1
+        perf_batch_size = self._perf_config[
+            'batch-size'] if 'batch-size' in self._perf_config else 1
+
+        return max_batch_size >= perf_batch_size
```
Comment (nv-braf) [inline model_analyzer/config/run/model_run_config.py:82]: Add newline

File: model_analyzer/config/run/run_config.py [hunk 2]
```diff
@@ -15,86 +15,50 @@
 
 class RunConfig:
     """
-    A class that encapsulates all the information
-    needed to complete a single run of the Model
-    Analyzer. One run corresponds to
-    one ModelConfig and produces one ModelResult.
+    A class that encapsulates all the information needed to run one or more
```
Comment (nv-braf) [inline model_analyzer/config/run/run_config.py:18]: Remove "A class that"

File: model_analyzer/record/metrics_manager.py [hunk 3]
```diff
@@ -167,26 +167,29 @@ def profile_server(self):
 
     def execute_run_config(self, run_config):
         """
-        Executes the run config. Returns obtained measurement. Also sends 
+        Executes the RunConfig. Returns obtained measurement. Also sends 
         measurement to the result manager
         """
 
+        model_run_config = run_config.model_run_configs()[0]
```
Comment (nv-braf) [inline model_analyzer/record/metrics_manager.py:174]: Is this changing when we go to multi-model? And if so, add a comment here

File: model_analyzer/record/metrics_manager.py [hunk 4]
```diff
@@ -315,23 +318,25 @@ def profile_model(self, run_config):
...
-            run_config object corresponding to the model being profiled.
+            RunConfig object corresponding to the models being profiled.
 
         Returns
         -------
         (dict of lists, list)
             The gpu specific and non gpu metrics
         """
 
+        model_run_config = run_config.model_run_configs()[0]
```
Comment (nv-braf) [inline model_analyzer/record/metrics_manager.py:329]: Is this changing when we go to multi-model? And if so, add a comment here

File: tests/test_run_config_generator.py [hunk 5]
```diff
@@ -71,6 +72,286 @@ def test_default_config_single_model(self):
...
+
+        with patch.object(TestRunConfigGenerator,
+                          "_get_next_perf_throughput_value") as mock_method:
+            mock_method.side_effect = perf_throughput_values
+            self._run_and_test_run_config_generator(
+                yaml_content, expected_config_count=expected_num_of_configs)
+
+    def test_early_backoff_root_model(self):
+        """
+        Test the case where there are two models, and the 'root' model has early backoff
```
Comment (nv-braf) [inline tests/test_run_config_generator.py:261]: By root do you mean model1? Maybe say first model instead

File: tests/test_run_config_generator.py [hunk 6]
```diff
@@ -71,6 +72,286 @@ def test_default_config_single_model(self):
...
+
+        with patch.object(TestRunConfigGenerator,
+                          "_get_next_perf_throughput_value") as mock_method:
+            mock_method.side_effect = perf_throughput_values
+            self._run_and_test_run_config_generator(
+                yaml_content, expected_config_count=expected_num_of_configs)
+
+    def test_measurement_list(self):
+        """
+        Test that the root model gets a list of all measurements since the last time it took a step,
```
Comment (nv-braf) [inline tests/test_run_config_generator.py:308]: Same comment here

File: tests/test_run_config_generator.py [hunk 7]
```diff
@@ -92,23 +373,27 @@ def _run_and_test_run_config_generator(self, yaml_content,
...
+        rcg = RunConfigGenerator(config, config.profile_models, MagicMock())
 
         run_configs = []
         rcg_config_generator = rcg.next_config()
         while not rcg.is_done():
             run_configs.append(next(rcg_config_generator))
             rcg.set_last_results(self._get_next_fake_results())
 
-        self.assertEqual(expected_config_count, len(set(run_configs)))
+        DEBUG = 0
```
Comment (nv-braf) [inline tests/test_run_config_generator.py:384]: Get rid of the debug code

**Review by nv-braf — COMMENTED** [review]
This is great stuff!

File: model_analyzer/config/generate/model_run_config_generator.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+from .config_generator_interface import ConfigGeneratorInterface
+from .generator_factory import ConfigGeneratorFactory
+from .perf_analyzer_config_generator import PerfAnalyzerConfigGenerator
+
+from model_analyzer.config.run.model_run_config import ModelRunConfig
+
+
+class ModelRunConfigGenerator(ConfigGeneratorInterface):
+    """
+    Generates all ModelRunConfigs to execute given a model
```
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply model_analyzer/config/generate/model_run_config_generator.py]: updated to this:

    Given a model, generates all ModelRunConfigs (combination of 
    ModelConfig and PerfConfig)

File: model_analyzer/record/metrics_manager.py (hunk 3 above)
Comment (tgerdesnv) [reply model_analyzer/record/metrics_manager.py:175]: added comment

File: tests/test_run_config_generator.py (hunk 5 above)
Comment (tgerdesnv) [reply tests/test_run_config_generator.py]: I meant the first one called in the recursive generator stack. I've tried to clarify the comments

**Pull request merged**